### PR TITLE
Implement np.linalg.matrix_rank

### PR DESCRIPTION
**Pull request by ziyadedher**
Implements [`np.linalg.matrix_rank`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.matrix_rank.html). 

Honestly not sure what sorcery needs to be going on for someone to want to take the gradient through this operation, but here we are anyway.

Related to #1999

### CODE REVIEW

**Review by hawkinsp — DISMISSED** [review]

**Comment by ziyadedher** [comment]
Requested changes should be implemented now. Thanks!

File: jax/numpy/linalg.py [hunk 1]
```diff
@@ -61,6 +61,19 @@ def svd(a, full_matrices=True, compute_uv=True):
   return lax_linalg.svd(a, full_matrices, compute_uv)
 
 
+@_wraps(onp.linalg.matrix_rank)
+def matrix_rank(M, tol=None):
+  M = _promote_arg_dtypes(np.asarray(M))
+  if M.ndim > 2:
+    raise TypeError("array should have 2 or fewer dimensions")
+  if M.ndim < 2:
+    return np.any(M != 0).astype(np.int32)
```
Comment (ziyadedher) [inline jax/numpy/linalg.py]: Casting to `np.int32` can cause issues if we are testing for `dtype` in `tests/linalg_test::testMatrixRank`. To get around this we are not checking for `dtype` in that test.

File: jax/numpy/linalg.py [hunk 2]
```diff
@@ -100,6 +100,19 @@ def matrix_power(a, n):
...
+  S = svd(M, full_matrices=False, compute_uv=False)
+  if tol is None:
+    tol = S.max() * np.max(M.shape) * np.finfo(S.dtype).eps
+  return sum(S > tol)
```
Comment (shoyer) [inline jax/numpy/linalg.py]: This should be sure to use np.sum rather than the builtin sum

**Comment by ziyadedher** [comment]
@shoyer Done :+1:.

**Comment by shoyer** [comment]
Thanks!

**Pull request merged**